File: adapters/go-gin/bugfab/schemas_test.go
package bugfab

import (
	"encoding/json"
	"strings"
	"testing"
)

func TestBugReportCreate_DefaultsApplied(t *testing.T) {
	var b BugReportCreate
	b.ProtocolVersion = "0.1"
	b.Title = "Save fails"
	b.ClientTS = "2026-04-27T15:00:00Z"
	b.applyDefaults()
	if b.ReportType != "bug" {
		t.Fatalf("want default report_type=bug, got %q", b.ReportType)
	}
	if b.Severity != "medium" {
		t.Fatalf("want default severity=medium, got %q", b.Severity)
	}
	if b.Tags == nil {
		t.Fatalf("tags should default to non-nil empty slice")
	}
}

func TestBugReportCreate_Validate_MissingProtocolVersion(t *testing.T) {
	b := BugReportCreate{Title: "x", ClientTS: "now"}
	versionErr, fieldErrs := b.Validate()
	if versionErr != nil {
		t.Fatalf("missing version should be field error, not version error")
	}
	if len(fieldErrs) == 0 {
		t.Fatalf("expected field error for missing protocol_version")
	}
}

func TestBugReportCreate_Validate_UnsupportedProtocolVersion(t *testing.T) {
	b := BugReportCreate{ProtocolVersion: "9.9", Title: "x", ClientTS: "now"}
	versionErr, _ := b.Validate()
	if versionErr == nil {
		t.Fatalf("expected version error for 9.9")
	}
}

func TestBugReportCreate_Validate_RejectsUnknownSeverity(t *testing.T) {
	// The conformance suite has an explicit severity="urgent" rejection
	// test — silently coercing to "medium" would fail conformance.
	b := BugReportCreate{
		ProtocolVersion: "0.1",
		Title:           "x",
		ClientTS:        "now",
		Severity:        "urgent",
	}
	_, fieldErrs := b.Validate()
	if len(fieldErrs) == 0 {
		t.Fatalf("expected field error for severity=urgent")
	}
	if !strings.Contains(fieldErrs[0].Msg, "urgent") {
		t.Fatalf("severity error should mention offending value, got %q", fieldErrs[0].Msg)
	}
	if fieldErrs[0].Type != "value_error.enum" {
		t.Fatalf("want enum-error type, got %q", fieldErrs[0].Type)
	}
}

func TestBugReportCreate_Validate_AcceptsAllLockedSeverities(t *testing.T) {
	for _, s := range []string{"low", "medium", "high", "critical"} {
		b := BugReportCreate{
			ProtocolVersion: "0.1",
			Title:           "x",
			ClientTS:        "now",
			Severity:        s,
		}
		_, fieldErrs := b.Validate()
		if len(fieldErrs) != 0 {
			t.Fatalf("severity %q should validate, got errors %+v", s, fieldErrs)
		}
	}
}

func TestBugReportCreate_Validate_RejectsTitleOver200(t *testing.T) {
	b := BugReportCreate{
		ProtocolVersion: "0.1",
		Title:           strings.Repeat("x", 201),
		ClientTS:        "now",
	}
	_, fieldErrs := b.Validate()
	if len(fieldErrs) == 0 {
		t.Fatalf("expected field error for 201-char title")
	}
}

func TestBugReportCreate_Validate_RejectsReporterFieldOver256(t *testing.T) {
	b := BugReportCreate{
		ProtocolVersion: "0.1",
		Title:           "x",
		ClientTS:        "now",
		Reporter:        Reporter{Email: strings.Repeat("a", 257)},
	}
	_, fieldErrs := b.Validate()
	if len(fieldErrs) == 0 {
		t.Fatalf("expected field error for 257-char reporter.email")
	}
}

func TestStatusUpdate_Validate_RejectsUnknown(t *testing.T) {
	u := BugReportStatusUpdate{Status: "resolved"}
	errs := u.Validate()
	if len(errs) == 0 {
		t.Fatalf("status=resolved must be rejected on write")
	}
}

func TestStatusUpdate_Validate_AcceptsKnown(t *testing.T) {
	for _, s := range []string{"open", "investigating", "fixed", "closed"} {
		u := BugReportStatusUpdate{Status: s}
		if errs := u.Validate(); len(errs) != 0 {
			t.Fatalf("status %q should validate, got %+v", s, errs)
		}
	}
}

func TestBugReportContext_RoundtripPreservesExtras(t *testing.T) {
	// Extra keys MUST survive round-trip per the wire protocol's
	// "context allows extras" guarantee.
	in := `{"url":"https://x","custom_field":"keep-me","module":"checkout"}`
	var ctx BugReportContext
	if err := json.Unmarshal([]byte(in), &ctx); err != nil {
		t.Fatalf("unmarshal failed: %v", err)
	}
	if ctx.URL != "https://x" {
		t.Fatalf("URL not parsed: %q", ctx.URL)
	}
	if ctx.Extras["custom_field"] != "keep-me" {
		t.Fatalf("extra field not captured: %+v", ctx.Extras)
	}
	out, err := json.Marshal(ctx)
	if err != nil {
		t.Fatalf("marshal failed: %v", err)
	}
	var roundtrip map[string]interface{}
	if err := json.Unmarshal(out, &roundtrip); err != nil {
		t.Fatalf("roundtrip unmarshal failed: %v", err)
	}
	if roundtrip["custom_field"] != "keep-me" {
		t.Fatalf("extra field lost on marshal: %s", string(out))
	}
	if roundtrip["url"] != "https://x" {
		t.Fatalf("named field lost on marshal: %s", string(out))
	}
}

File: adapters/phoenix/test/bug_fab/schemas_test.exs
defmodule BugFab.SchemasTest do
  use ExUnit.Case, async: true

  alias BugFab.Schemas

  describe "validate_create/1" do
    test "accepts a valid v0.1 payload" do
      meta = %{
        "protocol_version" => "0.1",
        "title" => "Save button is unresponsive",
        "client_ts" => "2026-04-27T15:29:58-07:00",
        "severity" => "high",
        "tags" => ["regression"]
      }

      assert {:ok, normalized} = Schemas.validate_create(meta)
      assert normalized["title"] == "Save button is unresponsive"
      assert normalized["severity"] == "high"
      assert normalized["report_type"] == "bug"
      assert normalized["reporter"]["name"] == ""
    end

    test "rejects unknown severity with 422" do
      meta = %{
        "protocol_version" => "0.1",
        "title" => "Bug",
        "client_ts" => "2026-04-27T15:29:58-07:00",
        "severity" => "urgent"
      }

      assert {:error, {:schema, cs}} = Schemas.validate_create(meta)
      errors = Schemas.format_errors(cs)
      assert Enum.any?(errors, fn e -> e.loc == "severity" end)
    end

    test "rejects missing protocol_version" do
      meta = %{"title" => "x", "client_ts" => "t"}
      assert {:error, :unsupported_protocol_version} = Schemas.validate_create(meta)
    end

    test "rejects wrong protocol_version" do
      meta = %{
        "protocol_version" => "0.2",
        "title" => "x",
        "client_ts" => "t"
      }

      assert {:error, :unsupported_protocol_version} = Schemas.validate_create(meta)
    end

    test "rejects unknown report_type with 422" do
      meta = %{
        "protocol_version" => "0.1",
        "title" => "x",
        "client_ts" => "t",
        "report_type" => "wishlist"
      }

      assert {:error, {:schema, _}} = Schemas.validate_create(meta)
    end

    test "rejects missing title" do
      meta = %{"protocol_version" => "0.1", "client_ts" => "t"}
      assert {:error, {:schema, cs}} = Schemas.validate_create(meta)
      errors = Schemas.format_errors(cs)
      assert Enum.any?(errors, fn e -> e.loc == "title" end)
    end

    test "preserves extra context keys verbatim" do
      meta = %{
        "protocol_version" => "0.1",
        "title" => "x",
        "client_ts" => "t",
        "context" => %{
          "url" => "https://example.com",
          "custom_field" => "preserved",
          "deep" => %{"a" => 1}
        }
      }

      assert {:ok, normalized} = Schemas.validate_create(meta)
      assert normalized["context"]["custom_field"] == "preserved"
      assert normalized["context"]["deep"] == %{"a" => 1}
    end

    test "rejects reporter fields longer than 256 chars" do
      long = String.duplicate("a", 257)

      meta = %{
        "protocol_version" => "0.1",
        "title" => "x",
        "client_ts" => "t",
        "reporter" => %{"name" => long}
      }

      assert {:error, {:schema, _}} = Schemas.validate_create(meta)
    end
  end

  describe "validate_status_update/1" do
    test "accepts valid status values" do
      for s <- ~w(open investigating fixed closed) do
        assert {:ok, %{"status" => ^s}} = Schemas.validate_status_update(%{"status" => s})
      end
    end

    test "rejects unknown status values" do
      assert {:error, %Ecto.Changeset{}} =
               Schemas.validate_status_update(%{"status" => "resolved"})
    end

    test "rejects missing status" do
      assert {:error, %Ecto.Changeset{}} = Schemas.validate_status_update(%{})
    end
  end
end
